File: src/strategies/macd.ts
import { Strategy } from "@/classes/strategy";
import { IObj } from "@/utils/interfaces";
import { RSI_ONLY } from "./ce-sma";

export class MACD_ONLY extends Strategy {
    name: string = "MACD ONLY";
    desc: string = `Enters: macd > 0  \n  Exit:  macd < 0`;

    buyCond(row: IObj): boolean {
        return row.macd > row.signal && row.c < row.bb_lower;
    }

    sellCond(row: IObj): boolean {
        return row.macd < row.signal && row.c > row.bb_upper;
    }
}
export class MACD_EXT extends Strategy {
    name: string = "MACD EXT";
    desc: string = `Enters: macd > 0  \n  Exit:  macd < 0`;

    buyCond(row: IObj): boolean {
        return row.hist > 0 && row.c > row.o;
    }

    sellCond(row: IObj): boolean {
        return row.hist < 0 && row.c < row.o
    }
}


export class MA_ONLY extends Strategy {
    name: string = "MA_ONLY";
    desc: string = `Enter: sma20 >  sma50, Exit: oposite `;

    buyCond(row: IObj): boolean {
        return row.sma_20 > row.sma_50; //cond && smaDiff > diff
    }

    sellCond(row: IObj): boolean {
        return row.sma_20 < row.sma_50;
 
    }
}

export class MACD_MA extends Strategy {
    name: string = "MACD_MA";
    desc: string = `Enter: macd > 0 && sma20 >  sma50, Exit: oposite`;

    buyCond(row: IObj): boolean {
        return MACD_EXT.prototype.buyCond(row) || MA_ONLY.prototype.buyCond(row)
    }

    sellCond(row: IObj): boolean {
        return MACD_EXT.prototype.sellCond(row) || MA_ONLY.prototype.sellCond(row)
    }
}
export class MACD_HL_HA extends Strategy {
    name: string = "MACD_HL_HA";
    desc: string = `Enter: macd > 0 && sma20 >  sma50, Exit: oposite`;

    buyCond(row: IObj): boolean {
        return MACD_EXT.prototype.buyCond(row) && HL_HA.prototype.buyCond(row)
    }

    sellCond(row: IObj): boolean {
        return MACD_EXT.prototype.sellCond(row) && HL_HA.prototype.sellCond(row)
    }
}
export class SMA_EXT extends Strategy {
    name: string = "SMA EXT";
    desc: string = `Enter: sma20 >  sma50 && low is < 5% from o, Exit: oposite `;

    buyCond(row: IObj): boolean {
        return (
            row.sma_20 > row.sma_50 && row.c > row.o
            //((row.ha_o - row.ha_l) / row.ha_l) * 100 <= 5
        ); //cond && smaDiff > diff
    }

    sellCond(row: IObj): boolean {
        return (
            row.sma_20 < row.sma_50 && row.c < row.o
            //((row.ha_h - row.ha_o) / row.ha_o) * 100 <= 5
        );
    }
}

export class CE_ONLY extends Strategy {
    name: string = "CE_ONLY";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return row.buy_signal == 1;
    }

    sellCond(row: IObj): boolean {
        return row.sell_signal == 1;
    }
}

export class CE_MA extends Strategy {
    name: string = "CE_MA";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return CE_ONLY.prototype.buyCond(row) && MA_ONLY.prototype.buyCond(row);
    }

    sellCond(row: IObj): boolean {
        return (
            CE_ONLY.prototype.sellCond(row) && MA_ONLY.prototype.sellCond(row)
        );
    }
}
export class CE_MACD extends Strategy {
    name: string = "CE_MACD";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return CE_ONLY.prototype.buyCond(row) || MACD_EXT.prototype.buyCond(row);
    }

    sellCond(row: IObj): boolean {
        return (
            CE_ONLY.prototype.sellCond(row) || MACD_EXT.prototype.sellCond(row)
        );
    }
}
export class MA_RSI extends Strategy {
    name: string = "MA_RSI";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return MA_ONLY.prototype.buyCond(row) || RSI_ONLY.prototype.buyCond(row);
    }

    sellCond(row: IObj): boolean {
        return (
            MA_ONLY.prototype.sellCond(row) || RSI_ONLY.prototype.sellCond(row)
        );
    }
}
export class RITA extends Strategy {
    name: string = "RITA";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return  (CE_ONLY.prototype.buyCond(row) || RSI_ONLY.prototype.buyCond(row) || MA_ONLY.prototype.buyCond(row))// || RSI_ONLY.prototype.buyCond(row))
    }

    sellCond(row: IObj): boolean {
        return(
            (CE_ONLY.prototype.sellCond(row) || RSI_ONLY.prototype.sellCond(row) || MA_ONLY.prototype.sellCond(row))// || RSI_ONLY.prototype.sellCond(row)
        );
    }
}

export class HL extends Strategy {
    name: string = "HL";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return row.c > row.o;
    }

    sellCond(row: IObj): boolean {
        return row.c < row.o;
    }
}
export class HL_HA extends Strategy {
    name: string = "HL_HA";
    desc: string = "JUST A CE";
    buyCond(row: IObj): boolean {
        return row.ha_c > row.ha_o// || RSI_ONLY.prototype.buyCond(row)
    }

    sellCond(row: IObj): boolean {
        return row.ha_c < row.ha_o// || RSI_ONLY.prototype.sellCond(row)
    }
}

export const strategies = [
    new MACD_ONLY(),
    new MACD_EXT(),
    new MACD_MA(),
    new MA_ONLY(),
    new SMA_EXT(),
    new CE_ONLY(),
    new CE_MA(),
    new CE_MACD(),
    new MA_RSI(),
    new RITA(),
    new HL(), new HL_HA(), new MACD_HL_HA()
];

File: strategies/macd.py


from classes.index import Strategy
from strategies.ce_sma import RSI_ONLY


class MACD_ONLY(Strategy):
    def __init__(self):
        super().__init__("MACD ONLY", "Enters: macd > 0\nExit: macd < 0")

    def buy_cond(self, row):
        return row['macd'] > row['signal'] and row['c'] < row['bb_lower']

    def sell_cond(self, row):
        return row['macd'] < row['signal'] and row['c'] > row['bb_upper']

class MACD_EXT(Strategy):
    def __init__(self):
        super().__init__("MACD EXT", "Enters: macd > 0\nExit: macd < 0")

    def buy_cond(self, row):
        return row['hist'] > 0 and row['c'] > row['o']

    def sell_cond(self, row):
        return row['hist'] < 0 and row['c'] < row['o']

class MA_ONLY(Strategy):
    def __init__(self):
        super().__init__("MA_ONLY", "Enter: sma20 > sma50, Exit: opposite")

    def buy_cond(self, row):
        return row['sma_20'] > row['sma_50']

    def sell_cond(self, row):
        return row['sma_20'] < row['sma_50']

class MACD_MA(Strategy):
    def __init__(self):
        super().__init__("MACD_MA", "Enter: macd > 0 && sma20 > sma50, Exit: opposite")

    def buy_cond(self, row):
        return MACD_EXT().buy_cond(row) or MA_ONLY().buy_cond(row)

    def sell_cond(self, row):
        return MACD_EXT().sell_cond(row) or MA_ONLY().sell_cond(row)

class SMA_EXT(Strategy):
    def __init__(self):
        super().__init__("SMA EXT", "Enter: sma20 > sma50 && low is < 5% from o, Exit: opposite")

    def buy_cond(self, row):
        return row['sma_20'] > row['sma_50'] and row['c'] > row['o']

    def sell_cond(self, row):
        return row['sma_20'] < row['sma_50'] and row['c'] < row['o']

class CE_ONLY(Strategy):
    def __init__(self):
        super().__init__("CE_ONLY", "JUST A CE")

    def buy_cond(self, row):
        return row['buy_signal'] == 1

    def sell_cond(self, row):
        return row['sell_signal'] == 1

class CE_MA(Strategy):
    def __init__(self):
        super().__init__("CE_MA", "JUST A CE")

    def buy_cond(self, row):
        return CE_ONLY().buy_cond(row) and MA_ONLY().buy_cond(row)

    def sell_cond(self, row):
        return CE_ONLY().sell_cond(row) and MA_ONLY().sell_cond(row)

class CE_MACD(Strategy):
    def __init__(self):
        super().__init__("CE_MACD", "JUST A CE")

    def buy_cond(self, row):
        return CE_ONLY().buy_cond(row) or MACD_EXT().buy_cond(row)

    def sell_cond(self, row):
        return CE_ONLY().sell_cond(row) or MACD_EXT().sell_cond(row)

class MA_RSI(Strategy):
    def __init__(self):
        super().__init__("MA_RSI", "JUST A CE")

    def buy_cond(self, row):
        return MA_ONLY().buy_cond(row) or RSI_ONLY().buy_cond(row)

    def sell_cond(self, row):
        return MA_ONLY().sell_cond(row) or RSI_ONLY().sell_cond(row)

class RITA(Strategy):
    def __init__(self):
        super().__init__("RITA", "JUST A CE")

    def buy_cond(self, row):
        return CE_ONLY().buy_cond(row) or RSI_ONLY().buy_cond(row) or MA_ONLY().buy_cond(row)

    def sell_cond(self, row):
        return CE_ONLY().sell_cond(row) or RSI_ONLY().sell_cond(row) or MA_ONLY().sell_cond(row)

class HL(Strategy):
    def __init__(self):
        super().__init__("HL", "JUST A CE")

    def buy_cond(self, row):
        return row['c'] > row['o']

    def sell_cond(self, row):
        return row['c'] < row['o']

class HL_HA(Strategy):
    def __init__(self):
        super().__init__("HL_HA", "JUST A CE")

    def buy_cond(self, row):
        print(row)
        return row['ha_c'] > row['ha_o']

    def sell_cond(self, row):
        return row['ha_c'] < row['ha_o']

strategies = [
    MACD_ONLY(),
    MACD_EXT(),
    MACD_MA(),
    MA_ONLY(),
    SMA_EXT(),
    CE_ONLY(),
    CE_MA(),
    CE_MACD(),
    MA_RSI(),
    RITA(),
    HL(),
    HL_HA()
]
